packages/@react-stately/table/src/useTableColumnResizeState.ts

```diff
@@ -0,0 +1,181 @@
+
+import {ColumnProps} from '@react-types/table';
+import {getContentWidth, getDynamicColumnWidths, getMaxWidth, getMinWidth, isStatic, mapToArray, parseStaticWidth} from './utils';
+import {GridNode} from '@react-types/grid';
+import {Key, MutableRefObject, useCallback, useRef, useState} from 'react';
+import {useLayoutEffect} from '@react-aria/utils';
+
+export interface ColumnResizeState<T> {
+  /** A ref whose current value is the state of all the column widths. */
+  columnWidths: MutableRefObject<Map<Key, number>>,
+  /** Setter for the table width. */
+  setTableWidth: (width: number) => void,
+  /** Trigger a resize and recalc. */
+  onColumnResize: (column: GridNode<T>, width: number) => void,
+  /** Callback for when onColumnResize has started. */
+  onColumnResizeStart: () => void,
+    /** Callback for when onColumnResize has ended. */
+  onColumnResizeEnd: (column: GridNode<T>, width: number) => void,
+    /** Getter for column widths. */
+  getColumnWidth(key: Key): number
+}
+
+export interface ColumnResizeStateProps<T> {
+  /** Collection of existing columns. */
+  columns: GridNode<T>[],
+  /** Callback to determine what the default width of a column should be. */
+  getDefaultWidth: (props) => string | number,
+  /** Callback that is invoked during the entirety of the resize event. */
+  onColumnResize?: (affectedColumnWidths: { key: Key, width: number }[]) => void,
+  /** Callback that is invoked when the resize event is ended. */
+  onColumnResizeEnd?: (affectedColumnWidths: { key: Key, width: number }[]) => void
+}
+
+export default function useTableColumnResizeState<T>(props: ColumnResizeStateProps<T>): ColumnResizeState<T> {
+
+  const {columns, getDefaultWidth} = props;
+  const columnsRef = useRef<GridNode<T>[]>();
+  const tableWidth = useRef<number>(null);
+  const isResizing = useRef<boolean>(null);
+  const startResizeContentWidth = useRef<number>();
+
+  const [columnWidths, setColumnWidths] = useState<Map<Key, number>>(new Map(columns.map(col => [col.key, 0])));
+  const columnWidthsRef = useRef<Map<Key, number>>(columnWidths);
+  const [resizedColumns, setResizedColumns] = useState<Set<Key>>(new Set());
+  const resizedColumnsRef = useRef<Set<Key>>(resizedColumns);
+
+  let getRealColumnWidth = useCallback((column: GridNode<T>) => {
+    let props = column.props as ColumnProps<T>;
+    return resizedColumns?.has(column.key) ? columnWidthsRef.current.get(column.key) : props.width ?? props.defaultWidth ?? getDefaultWidth?.(column.props) ?? '1fr';
+  }, [getDefaultWidth, resizedColumns]);
```

Comment:
```suggestion
  }, [getDefaultWidth, resizedColumns, props.width]);
```
is this missing props.width?

Reply:
the dependency array es-lint rule gets mad at me if I add this because I think the function is still called but it doesn't need to regenerate the function because of a change in `props.width` or `props.defaultWidth`, right? If you think that the rule is wrong in this case then maybe we should talk about it so I can understand where this would break, so I can better understand useCallback

Reply:
o, i see what happened, i confused my props
there's the outer scope of props, and then there's the inner scope of props coming from column.props
lets rename `let props = column.props ...` to `let colProps` or something so I don't make that mistake again